### DISCUSSION TITLE
ImageMagick cannot convert multiple small size image files to PDF with OCP environment

### DISCUSSION THREAD

**Discussion by yqliving** [ImageMagick/ImageMagick#4189] [Help]
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am using the latest version of ImageMagick
- [x] I have searched [open](https://github.com/ImageMagick/ImageMagick/issues) and [closed](https://github.com/ImageMagick/ImageMagick/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description
ImageMagick is used in the OCP environment with `swap option off`. When multiple files are uploaded for PDF conversion, for example, two tiff files, the first one can be converted, but the second one failed with an exception. 

We found a [related discussion about memory limitations](https://askubuntu.com/questions/724346/imagemagick-convert-and-low-ram
). `But it is about large image file > 10MB. However, our testing files are small tiff files, only around 50KB for each. And the processing failed even only two tiff files uploaded.`

**Logging during processing**
```
"error": {
"timedOut": false,
"killed": false,
"code": null,
"signal": "SIGKILL"
},
"level": "error",
"message": "uncaughtException: Command failed: \nError: Command failed: \n at ChildProcess. (/var/www/app/current/node_modules/imagemagick/imagemagick.js:88:15)\n at emitTwo (events.js:126:13)\n at ChildProcess.emit (events.js:214:7)\n at maybeClose (internal/child_process.js:915:16)\n at Process.ChildProcess._handle.onexit (internal/child_process.js:209:5)",
"stack": "Error: Command failed: \n at ChildProcess. (/var/www/app/current/node_modules/imagemagick/imagemagick.js:88:15)\n at emitTwo (events.js:126:13)\n at ChildProcess.emit (events.js:214:7)\n at maybeClose (internal/child_process.js:915:16)\n at Process.ChildProcess._handle.onexit (internal/child_process.js:209:5)",
"exception": true,
"date": "Thu Sep 10 2020 22:01:34 GMT+0000 (UTC)"
}
```

### Steps to Reproduce
1. upload 2 tiff files for processing (single page, small size tiff files)
2. in each request, imageMagic is called to do image to PDF conversion
3. check the result of imageMagic

### System Configuration
OpenShift on IBM Cloud with swap option set to off

- ImageMagick version: ImageMagick.x86_64 0:6.9.10.68-3.el7
- Environment (Operating system, version and so on): OpenShift on IBM Cloud
- Additional information:

**Comment by urban-warrior** [comment]
The exception points to the Javascript environment, perhaps the process exceeded resource limits.  If you can reproduce the problem with ImageMagick from the command-line, let us know. Otherwise assume the problem is within some other part of your workflow.

You could try setting your security policy, policy.xml, so that all pixels are pushed to disk rather than memory.  See if that makes a difference.  See https://imagemagick.org/script/security-policy.php and set the `area` limit to 0.

**Comment by cyshen1** [comment]
Thanks! We are currently on ImageMagick.x86_64 0:6.9.10.68-3.el7. To take the advance of using disk instead of memory, we need to upgrade to 7.0.7-0, am I right?

**Comment by urban-warrior** [comment]
For IM version 6, see https://legacy.imagemagick.org/script/security-policy.php.

**Comment by rferin** [comment]
Hello urban-warrior,
We have set the area to 0 but we are still getting this error when we try to process multiple images (one image works)

We are converting an 8.5"x11" TIFF at 300 dpi.

  "message": "uncaughtException: Command failed: convert: unable to write pixel cache `/tmp/magick-121g4TNPXfVxKGc': Cannot allocate memory @ error/cache.c/WritePixelCacheIndexes/5821.\n\nError: Command failed: convert: unable to write pixel cache `/tmp/magick-121g4TNPXfVxKGc': Cannot allocate memory @ error/cache.c/WritePixelCacheIndexes/5821.\n\n    at ChildProcess.<anonymous> (/var/www/app/current/node_modules/imagemagick/imagemagick.js:88:15)\n    at emitTwo (events.js:126:13)\n    at ChildProcess.emit (events.js:214:7)\n    at maybeClose (internal/child_process.js:925:16)\n    at Socket.stream.socket.on (internal/child_process.js:346:11)\n    at emitOne (events.js:116:13)\n    at Socket.emit (events.js:211:7)\n    at Pipe._handle.close [as _onclose] (net.js:554:12)",

"stack": "Error: Command failed: convert: unable to write pixel cache `/tmp/magick-121g4TNPXfVxKGc': Cannot allocate memory @ error/cache.c/WritePixelCacheIndexes/5821.\n\n    at ChildProcess.<anonymous> (/var/www/app/current/node_modules/imagemagick/imagemagick.js:88:15)\n    at emitTwo (events.js:126:13)\n    at ChildProcess.emit (events.js:214:7)\n    at maybeClose (internal/child_process.js:925:16)\n    at Socket.stream.socket.on (internal/child_process.js:346:11)\n    at emitOne (events.js:116:13)\n    at Socket.emit (events.js:211:7)\n    at Pipe._handle.close [as _onclose] (net.js:554:12)",

We have the policy.xml running with default values with the exception of area.

Are there more settings we can configure to have ImageMagick use disk?  I have not found a lot of documentation on the "disk" setting.  It looks like it defaults to 1 GB.   Could increasing this value improve this situation?  I tried increasing it to value="10GiB", but it did not resolve the issue.

Will any of the other settings in here also help?

I have not found documentation that describes the qualifiers in the profile.xml.  For example "GiB".  Is there a list of valid values for these settings?  Would GB also be valid?  It is not clear if they are optional, or are required, or if they can be changed.

Would you expect upgrading to ImageMagick 7 help the problem?

Thank you.

**Comment by rferin** [comment]
Hello,
We are still having a problem converting an image to PDF with ImageMagick.  It works when there are only a few images converting at a time.  Once we have more images converting to PDF at the same time, then the "unable to write pixel cache" exception noted above occurs.

I have tried setting the following in the policy.xml in ImageMagick6:

```
<policy domain="resource" name="area" value="0"/>
 <policy domain="resource" name="disk" value="10GiB"/> 
<policy domain="resource" name="thread" value="30"/> 
<policy domain="resource" name="file" value="2000"/> 
<policy domain="system" name="max-memory-request" value="0MiB"/> 
<policy domain="resource" name="memory" value="0GiB"/> 
```

The documentation for these settings is very terse.  I could not find documentation for every setting in the policy.xml.  

I have also tried setting equivalent environment variables that are documented.  That also did not help.  If both are set, which takes precedence, the policy XML or environment variable?  I could not find this documented either.

We are not allowed to enable disk swap on the K8 machine.  I also found that when running on my mac in docker, I get a similar error when processing a 50 page multi-page TIFF when converting it to PDF.  If I do just one TIF at a time, the conversion is successful.  If I try multiple copies of the TIF, then the memory error occurs.

I have switched the code so it calls the command line "convert" instead of calling the SDK "convert".  The command line convert also fails.  The exception returned is below:

`"message": "uncaughtException: Command failed: convert /tmp/bacabackend6ae69d20-0801-11eb-b9ab-05d6c09be809105MnMCNddjSTXq/prefix-105A2bIhZ17GGOf.tmp /var/storage/backend/prefix-105Ca5RajRo5OUx.pdf > /dev/null 2>&1\n\nError: Command failed: convert /tmp/bacabackend6ae69d20-0801-11eb-b9ab-05d6c09be809105MnMCNddjSTXq/prefix-105A2bIhZ17GGOf.tmp /var/storage/backend/prefix-105Ca5RajRo5OUx.pdf > /dev/null 2>&1\n\n    at ChildProcess.exithandler (child_process.js:275:12)\n    at emitTwo (events.js:126:13)\n    at ChildProcess.emit (events.js:214:7)\n    at maybeClose (internal/child_process.js:925:16)\n    at Process.ChildProcess._handle.onexit (internal/child_process.js:209:5)",`

What can we do to force imageMagick to use the disk instead of generating an out of memory condition?  Thank you.

**Comment by urban-warrior** [comment]
You likely want to set  `registry:temporary-path`.  See https://imagemagick.org/script/architecture.php#tera-pixel.

**Comment by rferin** [comment]
Hello, I have made several settings and ImageMagick is now using the disk.  These are the settings I have set:

```
<policy domain="resource" name="area" value="0"/>
<policy domain="resource" name="disk" value="10GiB"/> 
<policy domain="resource" name="thread" value="30"/> 
<policy domain="resource" name="file" value="2000"/> 
<policy domain="resource" name="temporary-path" value="/tmp"/> 
<policy domain="resource" name="memory" value="0GiB"/> 
<policy domain="system" name="max-memory-request" value="0MiB"/> 
```

The settings are preventing the out of memory errors that had been reported above.  However, the speed to convert a TIFF to PDF is much slower now.  While I would expect a slow down, it is slower than I would have expected.

For example, if I convert a 1 page b&w tif to PDF, when using default settings (not using disk), it takes 7 seconds.  When the disk is enabled, it takes 3m38s.  If I convert a 10 page multi-page TIF.  When not using disk, it takes 1m15s.  When configured to use disc it takes 1h, 3m for a single file.  

This is slower than I would have expected.  Is this kind of slow down expected?  Is there something I could do to improve the performance?

Ideally, I would  like ImageMagick to use memory until there is a low memory condition and then once there is a low memory situation, then start using disk.  That would cause the slow down only when a there is a low memory condition.  If we could specify a minimum free memory and if less than that, use the disk, that would be optimal.  As I understand the settings, they are limits for the images, so if you specify a memory size, IM will use memory if the image requires less and disk if the image requires more.  However, an out of memory error can still occur if the available memory is less than configured size.  At least, that is how I understand the settings.

Please let me know if there are some recommendations here to use memory until there is a low memory condition.

Also, we are still using V6.  Would this situation be improved if we were using V7?  Thanks.

**Comment by rferin** [comment]
Hello,

We are having trouble updating from ImageMagick v6 to v7.  We are installing on Centos 8.

We are installing the following components:
ImageMagick7-7.0.10.34-1.el8.remi.x86_64                                      
ImageMagick7-devel-7.0.10.34-1.el8.remi.x86_64                                
ImageMagick7-libs-7.0.10.34-1.el8.remi.x86_64 

For the PIP3 install we are specifying the following:
wand==0.6.3
magickwand==0.2

When starting our product, this error occurs:

File "/usr/local/lib/python3.6/site-packages/wand/color.py", line 10, in <module> from .api import library
File "/usr/local/lib/python3.6/site-packages/wand/api.py", line 177, in <module> 'Try to install:\n ' + msg)
 ImportError: MagickWand shared library not found.
You probably had not installed ImageMagick library.


We have not been able to resolve the error.  We have looked at the installation information page but we still have not resolved the issue.

Please confirm that we are installing the correct packages, if there is a package we should be installing that is missing, and provide any information that would help us resolve the issue.

Thank you.
